websites/Y/YouTube Music/presence.ts

```diff
@@ -2,90 +2,263 @@ const presence = new Presence({
-				Number(progressBar.getAttribute("value")),
-			watchID = document
-				.querySelector<HTMLAnchorElement>("a.ytp-title-link.yt-uix-sessionlink")
-				.href.match(/v=([^&#]{5,})/)[1],
-			presenceData: PresenceData = {
-				details: title,
-				state: [
-					navigator.mediaSession.metadata.artist,
-					navigator.mediaSession.metadata.album,
-				]
-					.filter(Boolean)
-					.join(" - "),
-				largeImageKey: cover
-					? navigator.mediaSession.metadata.artwork[0].src
-					: "ytm_lg",
-				smallImageKey: video.paused
+			presence.getSetting<boolean>("hidePaused"),
+			presence.getSetting<boolean>("browsing"),
+			presence.getSetting<boolean>("privacy"),
+		]),
+		{ mediaSession } = navigator,
+		watchID = document
+			.querySelector<HTMLAnchorElement>("a.ytp-title-link.yt-uix-sessionlink")
+			.href.match(/v=([^&#]{5,})/)?.[1],
+		repeatMode = document
+			.querySelector('ytmusic-player-bar[slot="player-bar"]')
+			.getAttribute("repeat-Mode_"),
+		videoElement =
+			document.querySelector<HTMLVideoElement>("video.video-stream");
+
+	if (videoElement) {
+		if (!videoListenerAttached) {
+			const videoListener = () => {
+				mediaTimestamps = presence.getTimestampsfromMedia(videoElement);
+				//* Don't ask me why the above function doesn't floor the end timestamp
+				mediaTimestamps[1] = Math.floor(mediaTimestamps[1]);
+			};
+
+			//* If video scrobbled, update timestamps
+			videoElement.addEventListener("seeked", videoListener);
+			//* If video resumes playing, update timestamps
+			videoElement.addEventListener("play", videoListener);
+
+			videoListenerAttached = true;
+		}
+		//* Element got removed from the DOM (eg, song with song/video switch)
+	} else {
+		prevTitleAuthor = "";
+		videoListenerAttached = false;
+	}
+
+	presenceData = null;
+
+	if (hidePaused && mediaSession.playbackState !== "playing")
+		return presence.clearActivity();
+
+	if (["playing", "paused"].includes(mediaSession.playbackState)) {
+		if (privacyMode) {
+			return mediaSession.playbackState === "playing"
+				? presence.setActivity({
+						largeImageKey: "ytm_lg",
+						details: "Listening to music",
+				  })
+				: presence.setActivity();
+		}
+
+		if (
+			!mediaSession.metadata?.title ||
+			//* Check if the video has a duration
+			isNaN(videoElement.duration)
+		)
+			return;
+
+		if (
+			prevTitleAuthor !==
+			mediaSession.metadata.title + mediaSession.metadata.artist
+		) {
+			mediaTimestamps = presence.getTimestampsfromMedia(videoElement);
+			//* Don't ask me why the above function doesn't floor the end timestamp
+			mediaTimestamps[1] = Math.floor(mediaTimestamps[1]);
+
+			prevTitleAuthor =
+				mediaSession.metadata.title + mediaSession.metadata.artist;
+		}
+
+		const albumArtistBtnLink = mediaSession.metadata.album
+				? [...document.querySelectorAll<HTMLAnchorElement>(".byline a")].at(-1)
+						?.href
+				: document.querySelector<HTMLAnchorElement>(".byline a")?.href,
+			buttons: [ButtonData, ButtonData?] = [
```

Comment:
Type isn't needed anymore

Reply:
Why is this not needed? Sometimes there are multiple author links and sometimes not and I can only show one author at a time

Reply:
The type is already declared https://github.com/PreMiD/Presences/blob/17cb9817a533349a0aab990197911b5f0eeed920/%40types/premid/index.d.ts#L41

Reply:
This is needed. Otherwise his type will be ButtonData[] and not [ButtonData, ButtonData?] which will make it incompatible when setting the data.